### PR TITLE
[AArch64] Stack probing for dynamic allocas in GlobalISel

### PR DESCRIPTION
**Pull request by momchil-velikov**

### CODE REVIEW

**Comment by aemerson** [comment]
This looks like too many changes for a single PR.

**Comment by momchil-velikov** [comment]
> This looks like too many changes for a single PR.

Yeah, it's too many, I have separate PRs for:
* https://github.com/llvm/llvm-project/pull/66524 (Stack probing in prologues)
* https://github.com/llvm/llvm-project/pull/66525 (Stack probing for dynamic allocation in SelectionDSG)
* this one (Stack probing for dynamic allocation in GlobalISel)
* maybe others

In the Phabricator days those would be dependent reviews, here they are just three branches on top of each other,
with each commit visible.

Perhaps I should mark the head of each PR in the commit message, so the structure is evident and people can start review from the first one, while still seeing the big picture.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 1]
```diff
@@ -8622,6 +8623,96 @@ bool AArch64InstrInfo::isReallyTriviallyReMaterializable(
...
+  MachineFunction::iterator MBBInsertPoint = std::next(MBB.getIterator());
+  MachineBasicBlock *LoopTestMBB =
+      MF.CreateMachineBasicBlock(MBB.getBasicBlock());
+  MF.insert(MBBInsertPoint, LoopTestMBB);
+  MachineBasicBlock *LoopBodyMBB =
+      MF.CreateMachineBasicBlock(MBB.getBasicBlock());
+  MF.insert(MBBInsertPoint, LoopBodyMBB);
+  MachineBasicBlock *ExitMBB = MF.CreateMachineBasicBlock(MBB.getBasicBlock());
+  MF.insert(MBBInsertPoint, ExitMBB);
+
```
Comment (oskarwirga) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:8648]:
```suggestion

    // Initialize flag indicating whether NZCV needs to be saved
  bool shouldSaveConditionFlag = false;
  bool foundStackAlloc = false;
  bool breakOuterLoop = false;
  Register SavedConditionReg;

  for (auto &currentBlock : MF) {
    for (auto instr = currentBlock.instr_begin(),
              end = currentBlock.instr_end();
         instr != end; ++instr) {
      if (instr->getOpcode() == AArch64::PROBED_STACKALLOC_DYN) {
        foundStackAlloc = true;
      }
      if (foundStackAlloc) {
        if (instr->definesRegister(llvm::AArch64::NZCV)) {
          shouldSaveConditionFlag = false;
          breakOuterLoop = true;
          break;
        }
        if (instr->readsRegister(llvm::AArch64::NZCV)) {
          shouldSaveConditionFlag = true;
          breakOuterLoop = true;
          break;
        }
      }
    }
    if (breakOuterLoop)
      break;
  }

  if (shouldSaveConditionFlag) {
    MachineRegisterInfo &MRI = MF.getRegInfo();

    const TargetRegisterClass *RC = &AArch64::GPR64RegClass;
    SavedConditionReg = MRI.createVirtualRegister(RC);

    // Define SavedConditionReg.
    BuildMI(MBB, MBBI, DL, TII->get(AArch64::ORRXrr), SavedConditionReg)
        .addReg(AArch64::XZR)
        .addReg(AArch64::XZR);
  }
  
```

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 2]
```diff
@@ -8622,6 +8623,96 @@ bool AArch64InstrInfo::isReallyTriviallyReMaterializable(
...
+    std::swap(Op1, Op2);
+    Cond = AArch64CC::GT;
+  }
+  BuildMI(*LoopTestMBB, LoopTestMBB->end(), DL, TII->get(AArch64::SUBSXrx64),
+          AArch64::XZR)
+      .addReg(Op1)
+      .addReg(Op2)
+      .addImm(AArch64_AM::getArithExtendImm(AArch64_AM::UXTX, 0))
+      .setMIFlags(MachineInstr::FrameSetup);
+
```
Comment (oskarwirga) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:8671]:
```suggestion

  if (shouldSaveConditionFlag) {
    // Save the NZCV flags to SavedConditionReg.
    BuildMI(*LoopTestMBB, LoopTestMBB->end(), DL, TII->get(AArch64::MRS))
        .addReg(SavedConditionReg)
        .addImm(AArch64::NZCV);
  }


```

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 3]
```diff
@@ -8622,6 +8623,96 @@ bool AArch64InstrInfo::isReallyTriviallyReMaterializable(
...
+  // LoopExit:
+  //   STR XZR, [TargetReg]
+  BuildMI(*ExitMBB, ExitMBB->begin(), DL, TII->get(AArch64::STRXui))
+      .addReg(AArch64::XZR)
+      .addReg(TargetReg)
+      .addImm(0)
+      .setMIFlags(MachineInstr::FrameSetup);
+
```
Comment (oskarwirga) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:8697]:
```suggestion


  if (shouldSaveConditionFlag) {
    // Restore the NZCV flags from SavedConditionReg.
    BuildMI(*ExitMBB, ExitMBB->end(), DL, TII->get(AArch64::MSR))
        .addImm(AArch64SysReg::NZCV)
        .addReg(SavedConditionReg)
        .addReg(AArch64::NZCV, RegState::Implicit | RegState::Define);
  }


```

**Review by oskarwirga — COMMENTED** [review]
I found an interaction between -fstack-clash-protection and -ftrivial-auto-var-init=zero that can be triggered with:
```
// RUN: %clang_cc1 "-triple" "aarch64-none-linux-android28" "-emit-obj" "-Os" "-fstack-clash-protection" "-mstack-probe-size=65536" "-ftrivial-auto-var-init=zero" "-vectorize-loops" "-x" "c" "%s" "-o" "/dev/null"

// This test case examines a specific interaction between the -fstack-clash-protection
// mitigation and the -ftrivial-auto-var-init=zero option.
// The -fstack-clash-protection option inserts a stack probe loop into the compiled code to protect
// against stack clashes. The -ftrivial-auto-var-init=zero option initializes automatic variables to zero.
// The test case was reduced from an instance where the compiler crashed when these options were used together.
// The crash occurred when condition flags were situated at the insertion point of the stack probe loop.

# 4 "" 3
typedef a;
typedef b;
typedef struct {
  b c
} d;
#define e(f, g, h) h f[g]
i(d *j) {
  a k;
  e(l, j->c == 2, a);
  if (j->c == 2)
    k = l;
  m(k);
}
```

I have a fix but I do not know if it is the correct way to do it. I will add it inline.

**Comment by momchil-velikov** [comment]
> I found an interaction between -fstack-clash-protection and -ftrivial-auto-var-init=zero that can be triggered with:

Thank you, I'll take a look.

**Comment by momchil-velikov** [comment]
Indeed, that's a bug in https://github.com/llvm/llvm-project/pull/67123/commits/86e003feef19b859736476b5b3f15e3637da7307 , specifically in the definition on `PROBED_STACKALLOC_DYN` pseudo-instruction - it must declare that it implicitly modifies flags.

Thank you again, I will fix that in a follow-up update.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp (hunk 1 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:9428]: I have now fixed the issue in the latest update (specifically with this patch https://github.com/llvm/llvm-project/pull/66524/commits/6f714da5566e68c4a02eaac45dabb0244d46ee20 )

Thanks!

**Review by oskarwirga — APPROVED** [review]

**Pull request merged**